### PR TITLE
ISSUE-1619: Implement session flow control using new Proton window API

### PR DESCRIPTION
**Pull request by kgiusti**

### CODE REVIEW

File: src/adaptors/amqp/container.c [hunk 1]
```diff
@@ -69,15 +69,34 @@ struct qd_session_t {
...
+// A session incoming window determines how many incoming frames the session will accept across all incoming links. This
+// places a limit on the number of incoming data bytes that have to be buffered on the session (connection max-frame *
+// max incoming window frames). The local session incoming window configuration is determined by the maxFrameSize and
+// maxSessionFrames configuration attributes of an AMQP listener/connector.
+//
+// The remote peers session window must be honored when writing output to a sending link. In addition we limit the
+// amount of outgoing data that can be buffered on a session before control is returned to Proton. This is necessary to
+// improve latency and allow capacity sharing among all links on the session.
+//
+const size_t   qd_session_max_outgoing_bytes = 1048576;   // max buffered bytes on a session
```
Comment (ganeshmurthy) [inline src/adaptors/amqp/container.c:86]:
```suggestion
const size_t   qd_session_max_outgoing_bytes = 1048576;   // 1 MB max bytes will be written to a session before backing off 
```

File: src/adaptors/amqp/container.c [hunk 2]
```diff
@@ -69,15 +69,34 @@ struct qd_session_t {
...
+const size_t   qd_session_low_outgoing_bytes = 524288;    // low watermark for max buffered bytes
```
Comment (ganeshmurthy) [inline src/adaptors/amqp/container.c:87]:
```suggestion
const size_t   qd_session_low_outgoing_bytes = 524288;    // Start writing outgoing bytes to a session only if the session outgoing capacity is greater than 0.5 MB
```

File: src/adaptors/amqp/container.c [hunk 3]
```diff
@@ -69,15 +69,34 @@ struct qd_session_t {
...
 
-    size_t  out_window_limit;
-    size_t  out_window_low_watermark;
-};
+const size_t qd_session_max_in_win_user    = (size_t) 8388608;    // AMQP application in window max bytes 8MB
```
Comment (ganeshmurthy) [inline src/adaptors/amqp/container.c:89]: > AMQP application

It is not clear to me what AMQP application means

File: src/policy.c [hunk 4]
```diff
@@ -652,17 +652,23 @@ bool qd_policy_approve_amqp_session(pn_session_t *ssn, qd_connection_t *qd_conn)
...
+    if (qd_conn->policy_settings) {
+        const qd_policy_spec_t *spec = &qd_conn->policy_settings->spec;
+        if (!spec->outgoingConnection && spec->maxSessionWindow) {
+            // Policy configures the window *in bytes* but Proton uses *frames*.  Convert to frames
+            uint32_t max_frame = spec->maxFrameSize ? spec->maxFrameSize : cf->max_frame_size;
+            *in_window = spec->maxSessionWindow / max_frame;
+            if (*in_window < 2)
+                *in_window = 2;
+        }
     }
```
Comment (ganeshmurthy) [inline src/policy.c:671]:
```suggestion

        if (qd_conn->policy_settings) {
        const qd_policy_spec_t *spec = &qd_conn->policy_settings->spec;
        if (!spec->outgoingConnection && spec->maxSessionWindow) {
            // Policy configures the window *in bytes* but Proton uses *frames*.  Convert to frames
            uint32_t max_frame = spec->maxFrameSize ? spec->maxFrameSize : cf->max_frame_size;
            *in_window = spec->maxSessionWindow / max_frame;
            if (*in_window < 2)
                *in_window = 2;
        } else {
              *in_window = 2;
        }
        
    } else {
              // Assume defaults will be used
            *in_window = cf->session_max_in_window;
    }
```

File: include/qpid/dispatch/amqp_adaptor.h [hunk 5]
```diff
@@ -34,7 +34,13 @@ typedef struct pn_link_t    pn_link_t;
 // These types are private to the AMQP adaptor:
 typedef struct qd_connection_t qd_connection_t;
 typedef struct qd_link_t       qd_link_t;
+typedef struct qd_session_t    qd_session_t;
 
+// Session windowing limits
+extern const size_t qd_session_max_in_win_user;   // incoming window byte limit for user connections
```
Comment (ganeshmurthy) [inline include/qpid/dispatch/amqp_adaptor.h:40]:
```suggestion
extern const size_t qd_session_max_incoming_win_user;   // incoming window byte limit for user connections
```

File: include/qpid/dispatch/amqp_adaptor.h [hunk 6]
```diff
@@ -34,7 +34,13 @@ typedef struct pn_link_t    pn_link_t;
...
+extern const size_t qd_session_max_in_win_trunk;  // incoming window byte limit for inter-router connections
```
Comment (ganeshmurthy) [inline include/qpid/dispatch/amqp_adaptor.h:41]:
```suggestion
extern const size_t qd_session_max_incoming_win_trunk;  // incoming window byte limit for inter-router connections
```

File: src/adaptors/amqp/container.c [hunk 7]
```diff
@@ -69,15 +69,34 @@ struct qd_session_t {
     sys_atomic_t    ref_count;
     pn_session_t   *pn_session;
     qd_link_list_t  q3_blocked_links;  ///< Q3 blocked if !empty
+    uint32_t remote_max_frame;
```
Comment (ganeshmurthy) [inline src/adaptors/amqp/container.c:72]:
```suggestion
    uint32_t remote_max_frame_size;
```

File: src/adaptors/amqp/server_config.c [hunk 8]
```diff
@@ -191,48 +189,49 @@ qd_error_t qd_server_config_load(qd_dispatch_t *qd, qd_server_config_t *config,
...
-    } else {
-        // Limited incoming frames.
-        if (is_64bit) {
-            // Specify this to proton by setting capacity to be
-            // the product (max_frame_size * ssn_frames).
-            config->incoming_capacity = (size_t)config->max_frame_size * (size_t)ssn_frames;
+    // Proton does not support maxSessions > 32768
+    int64_t value = (int64_t) qd_entity_get_long(entity, "maxSessions"); CHECK();
+    if (value == 0) {
+        value = 32768;  // default
```
Comment (ganeshmurthy) [inline src/adaptors/amqp/server_config.c:195]:
```suggestion
        value = QD_PN_MAX_SESSIONS;  // default
```
define QD_PN_MAX_SESSIONS in amqp.h ?

File: src/adaptors/amqp/server_config.h [hunk 9]
```diff
@@ -209,12 +209,14 @@ typedef struct qd_server_config_t {
...
-     * This calculation is done to satisfy proton pn_session_set_incoming_capacity().
+     * The session incoming window limit in frames. From maxSessionFrames configuration attribute
+     *
+     * The window indicates the maximum number of incoming *frames* that the session will buffer. This places a limit on
+     * the amount of memory the router needs to reserve to accomodate data arriving from the peer session endpoint.
+     *
+     * The maximum amount of memory is computed as (max_frame_size * session_max_in_window)
      */
-    size_t incoming_capacity;
+    uint32_t session_max_in_window;
```
Comment (ganeshmurthy) [inline src/adaptors/amqp/server_config.h:219]:
```suggestion
    uint32_t session_max_incoming_window;
```

File: src/adaptors/amqp/container.c [hunk 10]
```diff
@@ -69,15 +69,34 @@ struct qd_session_t {
...
+// improve latency and allow capacity sharing among all links on the session.
+//
+const size_t   qd_session_max_outgoing_bytes = 1048576;   // max buffered bytes on a session
+const size_t   qd_session_low_outgoing_bytes = 524288;    // low watermark for max buffered bytes
 
-    size_t  out_window_limit;
-    size_t  out_window_low_watermark;
-};
+const size_t qd_session_max_in_win_user    = (size_t) 8388608;    // AMQP application in window max bytes 8MB
+const size_t qd_session_max_in_win_trunk   = (size_t) 134217728;  // inter-router in window max bytes 128MB
```
[the remark this answers is not in the document]
Comment (kgiusti) [reply src/adaptors/amqp/container.c:90]: @ganeshmurthy but the maximum is not _per connection_, it is _per session_, right?  Hence the variable prefix "qd_session_..."

File: src/adaptors/amqp/container.c (hunk 3 above)
Comment (kgiusti) [reply src/adaptors/amqp/container.c:89]: @ganeshmurthy "user connection"?  Basically any connection via a listener with the role "normal". These would include connections from the control plane to do management stuff. All non router-2-router connections.

**Pull request merged**